scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/component/ComponentDefinitionExtractor.scala

```diff
@@ -73,35 +73,60 @@ object ComponentDefinitionExtractor {
-
+    def additionalConfigFromProvider(designerWideId: DesignerWideComponentId) = {
       additionalConfigsFromProvider
         .get(designerWideId)
         .map(ComponentAdditionalConfigConverter.toComponentConfig)
         .getOrElse(ComponentConfig.zero)
-        .copy(
-          componentId = Some(designerWideId)
-        )
     }
 
-    def withUiDefinitionForNotDisabledComponent[T](
+    def configFor(defaultConfig: ComponentConfig, withConfigFromProvider: Boolean) = {
+      val componentId                             = ComponentId(componentType, componentName)
+      val configFromAdditional                    = additionalConfigs.getConfig(componentId)
+      val combinedConfigWithoutConfigFromProvider = configFromAdditional |+| configFromDefinition |+| defaultConfig
+      val designerWideId =
+        combinedConfigWithoutConfigFromProvider.componentId.getOrElse(determineDesignerWideId(componentId))
+
+      val finalConfig = if (withConfigFromProvider) {
+        additionalConfigFromProvider(designerWideId) |+| combinedConfigWithoutConfigFromProvider
+      } else {
+        combinedConfigWithoutConfigFromProvider
+      }
+
+      finalConfig.copy(
+        componentId = Some(designerWideId)
+      )
+    }
+
+    def withUiDefinitionForNotDisabledComponent[T <: ComponentDefinitionWithImplementation](
         returnType: Option[TypingResult]
-    )(f: (ComponentUiDefinition, Map[ParameterName, ParameterConfig]) => T): Option[T] = {
+    )(toComponentDefinition: ResolvedComponentConfig => T): Option[T] = {
       val defaultConfig =
         DefaultComponentConfigDeterminer.forNotBuiltInComponentType(
           componentType,
           returnType.isDefined,
           customCanBeEnding
         )
-      val configFromAdditional                    = additionalConfigs.getConfig(componentId)
-      val combinedConfigWithoutConfigFromProvider = configFromAdditional |+| configFromDefinition |+| defaultConfig
-      val designerWideId                          = combinedConfigWithoutConfigFromProvider.componentId
-      val finalCombinedConfig = additionalConfigFromProvider(designerWideId) |+| combinedConfigWithoutConfigFromProvider
 
-      filterOutDisabledAndComputeFinalUiDefinition(finalCombinedConfig, additionalConfigs.groupName).map(f.tupled)
+      val finalEnrichedConfig = configFor(defaultConfig, withConfigFromProvider = true)
+      val finalRawConfig      = configFor(defaultConfig, withConfigFromProvider = false)
+
+      for {
+        enrichedComponentConfig <- filterOutDisabledAndComputeFinalUiDefinition(
+          finalEnrichedConfig,
+          additionalConfigs.groupName
+        )
+        rawComponentConfig <- filterOutDisabledAndComputeFinalUiDefinition(
+          finalRawConfig,
+          additionalConfigs.groupName
+        )
+      } yield toComponentDefinition(
+        ResolvedComponentConfig(
+          uiDefinition = enrichedComponentConfig._1,
```

Comment:
1. Let's use name variables in for comprehension instead of _1 and _2
2. What about `uiDefinition`? I think that you'll need to show things like `icon` or `docsUrl` before the additionalUiConfig application as well on your side. Even if you don't need this at this moment, IMO we should provide them for sake of consistency.